File: Depth-First Search/Binary tree/Find paths/LET_no437_Path Sum III.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
private:
    unordered_map<int, int> record;
    int res;
public:
    int pathSum(TreeNode* root, int targetSum) {
        record[0] = 1;
        res = 0;
        helper(root, targetSum, 0);
        return res;
    }
    void helper(TreeNode* root, int targetSum, int preSum) {
        if (!root)
            return;
        
        preSum += root->val;
        if (record.count(preSum - targetSum))
            res += record[preSum - targetSum];
        record[preSum] += 1;
        helper(root->left, targetSum, preSum);
        helper(root->right, targetSum, preSum);
        record[preSum] -= 1;
    }
};

File: Depth-First Search/Binary tree/Find paths/LET_no437_Path Sum III.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def pathSum(self, root: Optional[TreeNode], targetSum: int) -> int:
        record = collections.defaultdict(int)
        record[0] = 1
        self.res = 0
        self.helper(root, targetSum, record, 0)
        return self.res
    
    def helper(self, root, targetSum, record, preSum):
        if not root:
            return
        
        preSum += root.val
        self.res += record[preSum - targetSum]
        
        record[preSum] += 1
        self.helper(root.left, targetSum, record, preSum)
        self.helper(root.right, targetSum, record, preSum)
        record[preSum] -= 1